### ISSUE TITLE
improved object key types (-> key types with range can be described with JSONSchema )

### ISSUE THREAD

**Issue by craigphicks** [microsoft/TypeScript#42689]
# Suggestion

## 🔍 Search Terms

Generic Props

Key Types

## ✅ Viability Checklist

* [Y] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [Y] This wouldn't change the runtime behavior of existing JavaScript code
* [Y] This could be implemented without emitting different JS based on the types of the expressions
* [Y] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [Y] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

Particularly works towards goals 1 and 5
  - ~1 Statically identify constructs that are likely to be errors.
  - ~5 Produce a language that is composable and easy to reason about.

Moves away from non-goals 7
 - ~7 Introduce behaviour that is likely to surprise users. Instead have due consideration for patterns adopted by other commonly-used languages.

(... in my subjective opinion)

### proposal 2

Four new key types

 - `[key:js]`
     - Pass through for existing JS behavior, similarly to current `[key:string]', (but with changed usage rules)
 - `[key:safeint]`
    - apply same rule for  literal quoted, literal unquoted, number var, or string var:
       - `Number.isSafeInteger(Number(x)) && x.toString()===Number(x).toString()`
       -  Note: image of input under js-index-transformation is identity
 - `[key:nonnumber]`
    - apply same rule for  literal quoted, literal unquoted, number var, or string var:
       - `isNaN(Number(x))`
       -  Note: image of input under js-index-transformation is identity
 - `[key:safe]`
    - it is the union of `[key:safeint]` and `[key:nonnumber]`, for covenience.

Note that while `[key:safeint]` and  `[key:nonnumber]` are mutually exclusive, 
the union of `[key:safeint]` and  `[key:nonnumber]` is not `[key:js]` -  `the ugly numbers`, including floats,  are missing.
Think of it as having another key type `[key:uglynums]` which always has type `never`.

Usage constraints:
 - `[key:all]` may never be used with any of `[key:safeint]` or  `[key:nonnumber]` or `[key:safe]`
 - Any `[key:safeint]`, `[key:nonnumber]` or `[key:safe]`imply exclusion of 'the ugly numbers'
 - Any of `[key:safe]`, `[key:safeint]` and  `[key:nonnumber]` may be used together
 - Usage of `[key:string]` and `[key:number]` is allowed but deprecated, and neither of them may be used with new keys. 

If and when JS introduces `bigint` as a valid index than another typescript type can be introduced
  - `[key:bigint]`
     -  (typeof x==='bigint' || BigInt(x) doesn't throw) && BigInt(x).toString()===x.toString()  [+ allow trailing 'n']
     - cannot be used together with `[key:safeint]`

### Proposal 1
 
Just implement `[key:safe]` only.  That would be enough to remove strange non consecutive, and other strange keys like `Infinity` which are now included under  `[key:number]` 

## 💻 Use Cases

The point is to enable a constraint which is similar to `[key:number]:T`  (or its inverse image `[key:number]:never, [key:string]:T)`,
but constrains to values which can be represented in JSONSchema.

For example, In the case of `[key:number]:T`, `012345678901234568` is a member of the range of object indices satisfying `[key:number]` but `012345678901234567` is not.

Having a JSONSchema representable constraint is a worthy goal when  moving data parameters between JS/TS and non JS/TS environments.

**Comment by DanielRosenwasser** [comment]
This is something that doesn't really appear to be compositional or common enough that it warrants the complexity. It also touches on a bunch of other proposals like the `int` type, or negated types, which we've either rejected as out-of-scope or felt are too complex, so it especially doesn't feel like we should special-case them for index signatures.

**Comment by craigphicks** [comment]
@DanielRosenwasser - 

Well `bigint` is here so there is certainly no need for a general `int` type.  (Although they can't be used as indices in JS ... yet).

I can certainly understand prioritization, and key types are minor.

For the same reason (being minor), index-constraint checking is something that could often (perhaps even usually) be accomplished at runtime with low overhead.  That's probably a better way to approach it.

----

BTW I noticed a quirk in current usage of "excess keys signature tag".   I might as as well dump it here:

When "excess keys signature tag" is used , it enables some extra intellisense information about object intialising properties.
But the tag also means "allow extra properties".  These two objectives are uncorrelated - a user might want the feedback while NOT allowing extra properties (or visa versa).

Example with constraints and therefore intellisense, but extra keys are also allowed:
```
  const myObj1:{[key:string]:string,[key:number]:never} ={
    0o7:'TheSpyWho',  // error
    Infinity:'AndForver',
    101:'Dalmations',   // error
    2001:'Odyssey',    // error
  }
```

Example with extra keys not allowed (and therefore no intellisense):
```
  const myObj1 ={
    0o7:'TheSpyWho',  
    Infinity:'AndForver',
    101:'Dalmations',  
    2001:'Odyssey',    
  }
```
Hardly worth mentioning but I just did so anyway.
